### ISSUE TITLE
Request filtering mechanism in ASP.NET Core collector cannot work

### ISSUE THREAD

**Issue by jacostro** [open-telemetry/opentelemetry-dotnet#357]
Hi,
I recently tried to actually use a filtering delegate in `OpenTelemetry.Collector.AspNetCore` by providing a code which expected `HttpContext` instance as an argument. The delegate is used as an argument to subscribe to diagnostic source in `AspNetCoreCollector` constructor:

```csharp
public AspNetCoreCollector(ITracer tracer, AspNetCoreCollectorOptions options)
{
    this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(new HttpInListener("Microsoft.AspNetCore", tracer), options.EventFilter);
    this.diagnosticSourceSubscriber.Subscribe();
}
```

But, as it turns out, ASP.NET Core does not pass any argument to filtering delegate, which can be seen in its code:

```csharp
private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
{
    Activity activity = new Activity("Microsoft.AspNetCore.Hosting.HttpRequestIn");
    //
    // ... TL'DR ...
    //
    this._diagnosticListener.OnActivityImport(activity, (object) httpContext);
    if (this._diagnosticListener.IsEnabled("Microsoft.AspNetCore.Hosting.HttpRequestIn.Start"))
    {
        hasDiagnosticListener = true;
        this.StartActivity(activity, httpContext);
    }
    else
        activity.Start();
    return activity;
}
```

The filtering delegate is called by `_diagnosticListener.IsEnabled` with `null` arguments.

It seems that some another way of hooking filtering code is needed.

**Comment by jacostro** [comment]
OK, it's not exactly true that it cannot work, it just is not as easy to get information on current request as I expected. A possible workaround is to embed a reference to `IHttpContextAccessor` instance in the delegate - this is what I did. Still such wrapper could be provided by the collector code.

**Comment by lmolkova** [comment]
Here is where you have IsEnabled callback with HttpContext: 

https://github.com/aspnet/AspNetCore/blob/3b7cdc166aa3c9733fa60b56d91fc0fff9b11652/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L50

So you should get non-null first argument in **some** callbacks and it should have `HttpContext`. If this is not the case, this but probably belongs to ASP.NET Core.

**Comment by lmolkova** [comment]
Just checked if HttpContext is provided in the callback. Yes it does under `Microsoft.AspNetCore.Hosting.HttpRequestIn` name

![image](https://user-images.githubusercontent.com/2347409/70183278-554b6300-169a-11ea-9893-cd5a8ca4e9bc.png)

This is ASP.NET Core 2.1 app, the feature has been there from 2.0 and didn't change in 3.0. 

I'm closing this issue. If you feel I didn't address the concern - feel free to reopen.

**Comment by lmolkova** [comment]
Was able to repro it with ASDP.NET Core 3.

Caveat is that `loggingEnabled` is https://github.com/aspnet/AspNetCore/blob/3b7cdc166aa3c9733fa60b56d91fc0fff9b11652/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L50  is true and this was changed in 3.0.

**Comment by lmolkova** [comment]
fixed in #403
